csrc/multidevice/utils.cpp

```diff
@@ -96,4 +99,58 @@ bool isResharding(Expr* expr) {
   return !haveDifferentSharding(tv_ref, tvs).empty();
 }
 
+namespace {
+
+void shardAllLike(TensorView* ref, std::vector<TensorView*> tvs) {
+  for (auto tv : tvs) {
+    tv->setDeviceMesh(ref->getDeviceMesh());
+  }
+  scheduler_utils::parallelizeAllLike(ref, tvs, {ParallelType::DIDx});
+}
+
+void reshardBefore(Expr* expr, Fusion* fusion) {
+  NVF_ERROR(
+      expr->outputs().size() == 1,
+      "multi-output expressions are not supported");
+  NVF_ERROR(
+      expr->outputs().at(0)->isA<TensorView>(),
+      "the expression's output is not a TensorView");
+  TensorView* output = expr->outputs().at(0)->as<TensorView>();
+  std::unordered_set<TensorView*> inputs;
+  std::transform(
+      expr->inputs().begin(),
+      expr->inputs().end(),
+      std::inserter(inputs, inputs.end()),
+      [](Val* val) {
+        NVF_ERROR(
+            val->isA<TensorView>(),
+            "the expression's input is not a TensorView");
+        return val->as<TensorView>();
+      });
+  std::vector<TensorView*> new_inputs;
+  for (auto input : haveDifferentSharding(output, inputs)) {
+    // TODO: reuse cacheAfter?
+    // TODO: here we should add a mechanism to potentially reuse the inserted
+    // resharding accross all the consumer of the resharded tensor. This way we
+    // could avoid wasteful resharding set insertion.
+    TensorView* new_input = set(input);
```

Comment:
I think this will break for something like add(tv0, tv0) since it will try to reshard tv0 twice?

Reply:
Actually the scenario you mention is supported in the current version. I added this scenario in the test to support my claim.

Let me try to explain: in the situation you describe, where an Expr has twice the same input `tv0`, then:
1. in the surrounding for loop `auto input : haveDifferentSharding(output, inputs)`, `tv0` will be iterated on only once since `haveDifferentSharding` returns a set (without duplicates)
2. Then, `replaceValInExprInputs` will indeed substitute both "occurences" of `tv0` with the new resharded tv.